### PR TITLE
add requirements management policy

### PR DESCRIPTION
**Pull request by dhellmann**
This document describes the policies for adding and updating build and runtime dependencies of all InstructLab components.

### CODE REVIEW

**Comment by dhellmann** [comment]
FYI @n1hility @dhellmann @tiran @russellb @mairin

**Comment by tiran** [comment]
Several months ago, I experiemented with https://pypi.org/project/pip-compile-multi/ to generate a set of constraint files for our entire dependency chain. Yesterday I revived my old PR and tried again. pip-compile-multi breaks because packages like flash-attn have broken packaging metadata. Doug filed a ticket a while ago, but upstream hasn't agreed to accept the fix.

Related InstructLab ticket and PR to constraint versions:

- https://github.com/instructlab/instructlab/issues/1851
- https://github.com/instructlab/instructlab/pull/1863

**Comment by tiran** [comment]
The topic of version bounds came up in internal channels. The blog post https://iscinumpy.dev/post/bound-version-constraints/ explains very well how to deal with version constraints in Python. It links to several other blog posts about semver and version handling in Python ecosystem, too.

**Review by aakankshaduggal — APPROVED** [review]
Thanks @dhellmann!

File: docs/dependency-management.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+That maximum version setting should be used with care, though, and only when definitely needed, because it can prevent good new versions from being used and can cause incompatibility issues similar to pinning versions.
+
+Before setting a maximum version, wait for something to break, then set a cap to exclude the version that causes the breakage until the dependency is fixed or our code is updated to work with the new version of the dependency.
+Bad releases of dependencies (with incompatibilities or known bugs) can be excluded using the requirements syntax `!=a.b`.
+If the new dependency requires incompatible changes in our code, then it becomes the new minimum version for our requirements range.
+
+## Coordinating dependencies with downstream builders
+
+As the maintainers of InstructLab, it is our responsibility to ensure that the minimum versions of all of our dependencies are specified accurately so that downstream builders can know what to build.
+There are many reasons why a new version of a dependency available on PyPI.org might not be built downstream, including timing, source availability, build dependencies, and CVEs.
```
Comment (nathan-weinberg) [inline docs/dependency-management.md:41]:
```suggestion
There are many reasons why a new version of a dependency available on [PyPI](https://pypi.org/) might not be built downstream, including timing, source availability, build dependencies, and CVEs.
```

**Review by nathan-weinberg — COMMENTED** [review]
few nits but otherwise LGTM

**Review by nathan-weinberg — APPROVED** [review]

**Review by n1hility — APPROVED** [review]
sounds perfect to me

File: docs/dependency-management.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+Before setting a maximum version, wait for something to break, then set a cap to exclude the version that causes the breakage until the dependency is fixed or our code is updated to work with the new version of the dependency.
+Bad releases of dependencies (with incompatibilities or known bugs) can be excluded using the requirements syntax `!=a.b`.
+If the new dependency requires incompatible changes in our code, then it becomes the new minimum version for our requirements range.
+
+## Coordinating dependencies with downstream builders
+
+As the maintainers of InstructLab, it is our responsibility to ensure that the minimum versions of all of our dependencies are specified accurately so that downstream builders can know what to build.
+There are many reasons why a new version of a dependency available on [PyPI](https://pypi.org/) might not be built downstream, including timing, source availability, build dependencies, and CVEs.
+We cannot, therefore, assume that downstream builds reproduce the same set of packages that we get when we install using pip and PyPI
```
Comment (leseb) [inline docs/dependency-management.md:42]:
```suggestion
We cannot, therefore, assume that downstream builds reproduce the same set of packages that we get when we install using pip and PyPI.
```

**Review by leseb — COMMENTED** [review]
Great document, thanks! Do you think it would be beneficial to include a practice of documenting any changes to the dependency requirements? This could include the rationale for setting or changing version ranges, providing context for future maintainers and downstream builds? 

EDIT: "squash and merge" is not enabled in this repository so please do it locally and push again, thanks!

**Comment by dhellmann** [comment]
> Great document, thanks! Do you think it would be beneficial to include a practice of documenting any changes to the dependency requirements? This could include the rationale for setting or changing version ranges, providing context for future maintainers and downstream builds?

That sounds like a good follow-up for someone else to do, yes.

**Comment by dhellmann** [comment]
> EDIT: "squash and merge" is not enabled in this repository so please do it locally and push again, thanks!

Done

File: docs/dependency-management.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+4. Only apply "caps" to dependencies (using `<`) when that dependency has established a pattern of producing new releases with breaking changes.
+5. Pin versions in CI, with frequent automated updates.
+
+Best practices for python dependencies call for using ranges in package requirements and pinning versions only in CI jobs.
+
+Using pinned versions in, for example, `requirements.txt` or `constraints.txt`, allows us to know and advertise exactly what versions have been tested in CI.
+That information is useful for users and re-packagers to understand which versions of dependencies are compatible with more specificity than the ranges provide.
+Tools like Dependabot will submit PRs to automatically update those pins to help us keep up with new releases of all of our dependencies.
+
+Pinning to specific versions in the package dependencies in `pyproject.toml` or `requirements.txt` is not a good practice.
```
Comment (mairin) [inline docs/dependency-management.md:20]: in line 16, it suggests to use pinned versions in requirements.txt for CI, but here it suggests using pinned versions in requirement.txt is not a good practice. Just found this confusing.

Comment (dhellmann) [reply docs/dependency-management.md:20]: You don't want to pin your package dependencies. You do want to pin your CI packages. That's done in separate files. Both tend to be named requirements.txt. The earlier text had said `requirements.in` but that was changed to `requirements.txt`.

I don't know if you're using separate requirements/constraints lists for CI, yet.

Comment (nathan-weinberg) [reply docs/dependency-management.md:20]: I made that suggestion assuming it was a typo - if not feel free to change it back

Comment (dhellmann) [reply docs/dependency-management.md]: The point of the suggestions is getting lost in the concern about specific filenames, so I've dropped the filenames and focused on how the requirements are used in different cases.

If this part is still confusing, I can just remove it all. The most important part for me right now is that the team understands their responsibility for keeping minimum versions up to date.

**Pull request merged**